v1/src/main/java/com/google/cloud/teleport/templates/common/CsvConverters.java

```diff
@@ -0,0 +1,306 @@
+        order = 1,
+        optional = true,
+        description = "BigQuery Deadletter table to send failed inserts.",
+        helpText =
+            "Messages failed to reach the target for all kind of reasons (e.g., mismatched schema, malformed json) are written to this table.",
+        example = "your-project:your-dataset.your-table-name")
+    ValueProvider<String> getDeadletterTable();
+
+    void setDeadletterTable(ValueProvider<String> deadletterTable);
+
+    @TemplateParameter.Boolean(
+        order = 2,
+        optional = true,
+        description = "Input CSV files contain a header record.",
+        helpText =
+            "Input CSV files contain a header record (true/false). Only required if reading CSV files.")
+    @Default.Boolean(false)
+    ValueProvider<Boolean> getContainsHeaders();
+
+    void setContainsHeaders(ValueProvider<Boolean> containsHeaders);
+
+    @TemplateParameter.Text(
+        order = 3,
+        optional = true,
+        description = "Column delimiter of the data files.",
+        helpText =
+            "The column delimiter of the input text files. Default: use delimiter provided in csvFormat",
+        example = ",")
+    @Default.InstanceFactory(DelimiterFactory.class)
+    ValueProvider<String> getDelimiter();
+
+    void setDelimiter(ValueProvider<String> delimiter);
+
+    @TemplateParameter.Text(
+        order = 4,
+        optional = true,
+        description = "CSV Format to use for parsing records.",
+        helpText =
+            "CSV format specification to use for parsing records. Default is: Default. See https://commons.apache.org/proper/commons-csv/apidocs/org/apache/commons/csv/CSVFormat.html for more details. Must match format names exactly found at: "
+                + "https://commons.apache.org/proper/commons-csv/apidocs/org/apache/commons/csv/CSVFormat.Predefined.html")
+    @Default.String("Default")
+    ValueProvider<String> getCsvFormat();
+
+    void setCsvFormat(ValueProvider<String> csvFormat);
+
+    @TemplateParameter.Text(
+        order = 5,
+        optional = true,
+        regexes = {"^(US-ASCII|ISO-8859-1|UTF-8|UTF-16)$"},
+        description = "CSV file encoding",
+        helpText =
+            "CSV file character encoding format. Allowed Values are US-ASCII"
+                + ", ISO-8859-1, UTF-8, UTF-16")
+    @Default.String("UTF-8")
+    ValueProvider<String> getCsvFileEncoding();
+
+    void setCsvFileEncoding(ValueProvider<String> csvFileEncoding);
+  }
+
+  /**
+   * Default value factory to get delimiter from Csv format so that if the user does not pass one
+   * in, it matches the supplied {@link CsvPipelineOptions#getCsvFormat()}.
+   */
+  public static class DelimiterFactory implements DefaultValueFactory<ValueProvider<String>> {
+
+    @Override
+    public ValueProvider<String> create(PipelineOptions options) {
+      return DualInputNestedValueProvider.of(options.as(CsvPipelineOptions.class).getCsvFormat(),
```

Comment:
do we have a single input value provider to avoid the hack with the second input null here?

Reply:
NestedValueProvider.of?